### ISSUE TITLE
omxParallelCI not using OpenMP or multi-core

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#77]
As shown in #75,  CIs are not using parallel.

A no-brainer for multi-thread dispatch (or even multi-CPU), but what code issues in OpenMx do we need to solve to implement this (I had wrongly assumed it would come for free if each CI was put into a fake submodel and reassembled... but omxParallelCI timings show this is not the case: each runs sequentially.

**Comment by RMKirkpatrick** [comment]
I believe at one point (in the days of OpenMx v1.x), the plan was for OpenMx to be able to run independent submodels in parallel via 'snowfall' (i.e., across multiple physical CPUs).  That was never implemented.  But, `omxParallelCI()` got its name because it would have run each confidence interval in parallel if the API to 'snowfall' were ever implemented.

However, what makes you so sure that `omxParallelCI()` isn't using multiple threads via OpenMP?  It should be parallelizing over the rows of the dataset whenever it evaluates the -2logL, as that's the default behavior for FIML analyses.

**Comment by tbates** [comment]
Just watching processor use: 1-core. Will check again

**Comment by RMKirkpatrick** [comment]
Can this issue be closed?

**Comment by tbates** [comment]
I guess, if there's no roadmap for enabling omxParallelCI to use OpenMP?

**Comment by RMKirkpatrick** [comment]
You mean 'snowfall'?  `omxParallelCI()` invokes `mxRun()`, and the backend will use OpenMP if it has been compiled to do so, in order to (for instance) divvy up row-likelihood evaluations among multiple threads.

**Comment by tbates** [comment]
Yeah: snowfall. The help now says that parallel is not implemented. 

So issue could be closed if there's no plan to implement snowfall support for CIs on cluster/multi-CPU systems.

Curiously, at least for simple models, CIs run faster with cores set to 1 than to using all the available cores:

```R

m1 = umxRAM("CI", data = twinData,
	umxPath(c("wt1", "ht1"), to = "bmi1"),
	umxPath("wt1", with = "ht1"),
	umxPath(v.m. = c("wt1", "ht1", "bmi1"))
)

umx_set_cores(1)
m2 = umxConfint(m1, parm="all", run=TRUE)

umx_set_cores(4)
m3 = umxConfint(m1, parm="all", run=TRUE)

umx_time(c(m1, m2, m3))

noCI: 00.02 seconds, 
1coreCIs: 00.07 seconds(∆: 0.05), 292.006%
4coresCIs: 00.15 seconds(∆: 0.128), 591.056%

```

**Comment by khusmann** [comment]
> However, what makes you so sure that omxParallelCI() isn't using multiple threads via OpenMP? It should be parallelizing over the rows of the dataset whenever it evaluates the -2logL, as that's the default behavior for FIML analyses.

I don't think this is happening in the current build? As I mentioned in the other ticket before @tbates pointed me to this one, mxRun takes all cores to find solution, but then when I get to `[0/5] MxComputeGradientDescent(SLSQP) evaluations`... usage drops to a single core, which makes me think row-wise parallelization isn't happening. I'm running on a 20-core instance on an HPC, so the performance drop is pretty dramatic.

**Comment by mcneale** [comment]
Can we check the code base to see whether doing CIs one per core was started as a project, for which switching off the row-level parallelization would be necessary?  OTherwise, I figure evaluating the likelihood should proceed as usual, parallel over FIML rows wherever possible.  Or is the call to evaluate likelihood’s for CI purposes switching it off somehow?  I am inclined to believe the user that parallel is completely broken for CIs, which is not good.

**Comment by jpritikin** [comment]
Can somebody reproduce the problem with `mxOption(key="Parallel diagnostics", value = "Yes")` ?

**Comment by tbates** [comment]
```R
# set debug option
mxOption(key="Parallel diagnostics", value = "Yes") 
```

```R

m1 = umxRAM("test", data = twinData,
	umxPath(c("wt1", "ht1"), to = "bmi1"),
	umxPath("wt1", with = "ht1"),
	umxPath(v.m. = c("wt1", "ht1", "bmi1"))
)
```

```R
umx_set_cores(1)

m2 = umxConfint(m1, parm="all", run=TRUE)
m2 = mxRename(m2, "one_core")

```
> Running CI with 9 parameters
CI.fitfunction: rowwiseParallel = 1
FitContext::createChildren: max threads set to 1
FitContext::createChildren: max threads set to 1


```R
umx_set_cores(4)
m3 = umxConfint(m1, parm="all", run=TRUE)
m3 = mxRename(m3, "four_candles")
```
> Running CI with 9 parameters
CI.fitfunction: rowwiseParallel = 1
FitContext::createChildren: CI.fitfunction is an OpenMP user
FitContext::createChildren: create 4 FitContext for parallel processing; OpenMP user=1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
FitContext::createChildren: CI.fitfunction is an OpenMP user
FitContext::createChildren: create 4 FitContext for parallel processing; OpenMP user=1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
FitContext::createChildren: CI.fitfunction is an OpenMP user
FitContext::createChildren: create 4 FitContext for parallel processing; OpenMP user=1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
CI.fitfunction: rowwiseParallel = 1
...

```R
umx_time(c(m1, m2, m3))
```

> CI: 00.05 seconds, 
one_core: 00.42 seconds(Δ: 0.368), 722.751%
four_candles: 00.42 seconds(Δ: 0.368), 722.985%

No difference at all 1 to 4 cores for CIs... 2020 same as was in 2018

```R
# un-debug
mxOption(key="Parallel diagnostics", value = "No") 
```

**Comment by jpritikin** [comment]
Your test runs too fast on my machine to tell whether anything is different, but I suggest to turn off `rowwiseParallel`,

```
library(umx)

mxOption(key="Parallel diagnostics", value = "Yes")

m1 = umxRAM("test", data = twinData,
            umxPath(c("wt1", "ht1"), to = "bmi1"),
            umxPath("wt1", with = "ht1"),
            umxPath(v.m. = c("wt1", "ht1", "bmi1"))
)
m1$fitfunction$rowwiseParallel <- FALSE

umx_set_cores(4)
m3 = umxConfint(m1, parm="all", run=TRUE)
m3 = mxRename(m3, "four_candles")
```

**Comment by khusmann** [comment]
I get a speedup with @tbates' example:

```
test: 00.79 seconds, 
one_core: 00.44 seconds(Δ: -0.346), 56.457%
four_candles: 00.31 seconds(Δ: -0.475), 40.18%
```

In my other larger (ordinal) model, it looks like I was mistaken and the differential slowdown between fit and CI was due to things I added to debug my earlier openmp issue. I've created a clean build now with HEAD and ran with parallel diagnostics and get: `FitContext::createChildren: dsc_model_none.fitfunction cannot be duplicated` whenever it tries to fit with the parallel diagnostics enabled. It looks like it's spawning threads, but they all hover at 0-2% cpu. In an earlier revision I'm used to this blasting all my CPUs (e6b94c0e02f) and taking forever to solve, but now it's relatively quick even though I don't see it using much of the other cores (maybe this is from some of the ordinal stuff @jpritikin has worked on since then?)... so I guess that's good overall?

It might be nice to have a set of large models of different types we can use as examples / regression tests for parallel features.

**Comment by jpritikin** [comment]
> taking forever to solve, but now it's relatively quick

I'm glad you are happy with the performance, but

> It looks like it's spawning threads, but they all hover at 0-2% cpu.

That's not good. What kind of fitfunction are you using that cannot be duplicated?

> It might be nice to have a set of large models of different types we can use as examples / regression tests for parallel features.

Agreed. There are settings like `rowwiseParallel` that are probably not widely known or tested.

**Comment by tbates** [comment]
here's a slower model: some speedup for opening up more cores, but nothing like for running a model (about 30% faster with 4 cores compared to 1, when something more like > 300% would be expected).

```R
data(twinData)
twinData= umx_scale_wide_twin_data(data=twinData,varsToScale="wt",sep= "")
# Cut BMI column to form ordinal obesity variables
obLevels   = c('normal', 'overweight', 'obese')
cuts       = quantile(twinData[, "bmi1"], probs = c(.5, .2), na.rm = TRUE)
twinData$obese1=cut(twinData$bmi1,breaks=c(-Inf,cuts,Inf),labels=obLevels)
twinData$obese2=cut(twinData$bmi2,breaks=c(-Inf,cuts,Inf),labels=obLevels)
# Make the ordinal variables into mxFactors
ordDVs = c("obese1", "obese2")
twinData[, ordDVs] = umxFactor(twinData[, ordDVs])
mzData = twinData[twinData$zygosity %in% "MZFF",] 
dzData = twinData[twinData$zygosity %in% "DZFF",]
mzData = mzData[1:80,] # just top 80 so example runs in a couple of secs
dzData = dzData[1:80,]
m1 = umxACE(selDVs= c("wt","obese"), dzData= dzData, mzData= mzData, sep='')

m1$MZ$fitfunction$rowwiseParallel = FALSE
m1$DZ$fitfunction$rowwiseParallel = FALSE

umx_set_cores(1); m2 = umxConfint(m1, run=TRUE)
umx_set_cores(4); m3 = umxConfint(m1, run=TRUE)
umx_time(c(m1, m2,m3))

```
ACE: 00.36 seconds, 
ACE: 04.13 seconds(Δ: 3.763), 1122.037%
ACE: 03.08 seconds(Δ: 2.72), 838.862%


Also, this model returns mostly NA for CIs... I bet they are fine but being suppressed (but can't remember off hand where to look for the estimates inside the model that don't get printed as CIs)

> Running ACE with 11 parameters
               lbound estimate ubound lbound Code ubound Code                        
top.a_std[1,1]     NA    0.904     NA          NA          NA
top.a_std[2,1]     NA    0.579     NA          NA          NA
top.a_std[1,2]      0    0.000     NA           0          NA
top.a_std[2,2]     NA    0.343     NA          NA          NA
top.c_std[1,1]     NA    0.151     NA          NA          NA
top.c_std[2,1]     NA    0.465     NA          NA          NA
top.c_std[1,2]      0    0.000      0           0           0
top.c_std[2,2]     NA    0.000     NA          NA          NA
top.e_std[1,1]     NA    0.401     NA          NA          NA
top.e_std[2,1]     NA    0.540     NA          NA          NA
top.e_std[1,2]     NA    0.000      0          NA           0
top.e_std[2,2]     NA    0.197     NA          NA          NA

**Comment by khusmann** [comment]
> That's not good. What kind of fitfunction are you using that cannot be duplicated?

It's a MxFitFunctionMultigroup for the supermodel, and each submodel has MxFitFunctionWLS

**Comment by RMKirkpatrick** [comment]
Perhaps then this issue is specific to multigroup models?

**Comment by mcneale** [comment]
I am dismayed at how poorly the CIs are currently working.  I object to having NA printed for a CI if it is at or very very close to a parameter boundary (explicit for sure).  Should just print the value and an ! for a note.  I do not understand why codes are NA for a bunch of CI attempts.  If they don't succeed, they should have a code, why is it NA??  W

**Comment by jpritikin** [comment]
> I am dismayed at how poorly the CIs are currently working.

Then open a new issue. This issue is about threads + CIs.

**Comment by RMKirkpatrick** [comment]
So, does the multigroup fitfunction need to be modified to allow it to be duplicated, thereby allowing it to work with SLSQP's parallel numeric gradient elements?  Something similar had to be done for the GREML fitfunction (see b9167fe , and subsequently a7e1517 ).  One thing we would want to avoid is letting the multigroup function duplicate when one or more groups' fitfunctions have `rowwiseParallel=TRUE`.

Come to think of it, shouldn't the WLS fitfunction be allowed to duplicate as well, since it does not do any row-wise function evaluations?

**Comment by jpritikin** [comment]
@RMKirkpatrick Yeah, it looks like it could be a 1 liner, 88c2e46bd0388. Testing now.

**Comment by khusmann** [comment]
When I run @tbates's example with 20 cores and including the entire data set instead of just the first 80 rows (to really exercise things) I get:

```
ACE: 01.79 seconds, 
ACE: 06.78 seconds(Δ: 4.987), 377.536%
ACE: 26.70 seconds(Δ: 204.907), 11502.448%
```

When I re-enable rowwise parallel, this comes to:

```
ACE: 01.74 seconds, 
ACE: 06.76 seconds(Δ: 5.02), 387.337%
ACE: 08.56 seconds(Δ: 6.816), 490.114%
```

In all cases 20 cores is slower. Is this what we would expect? I still don't completely understand all the places where things can be parallelized. Why is disabling rowwise parallel making the confint calculation faster for @tbates but slower for me?

**Comment by RMKirkpatrick** [comment]
> In all cases 20 cores is slower. Is this what we would expect? I still don't completely understand all the places where things can be parallelized. Why is disabling rowwise parallel making the confint calculation faster for @tbates but slower for me?

@khusmann The answer to those questions depends on several relevant details.  Are you using WLS or ML?  How many free parameters do you have, vis-a-vis @tbates ?  How many rows of data do you have, vis-a-vis him?  How many logical CPUs (threads) does your processor support, compared to his?

With row-wise parallel, @mcneale did some benchmarking a few years ago, and found diminishing returns in performance above 11 threads or so.

**Comment by khusmann** [comment]
@RMKirkpatrick 

> Are you using WLS or ML?

I've copied @tbates code, which looks like it uses ML

> How many free parameters do you have, vis-a-vis @tbates ?

Same as @tbates 

> How many rows of data do you have, vis-a-vis him?

In his he caps both mzData and dzData it at 80 rows; I use the full dataset so 1232 and 751 rows, respectively.

> How many logical CPUs (threads) does your processor support, compared to his?

I'm running on an HPC node with 20 cores allocated to me. $OMP_NUM_THREADS is also correctly set to 20 in my env.

Where would the bottleneck be here? Isn't this a pretty ideal thing to parallelize?

Here's the code I'm running below. It's the same as @tbates code, only using the full dataset instead of just the top 80.

```
library(umx)                                                                                                                                                               
 data(twinData)                                                                                                                                                             
 mxOption(key="Parallel diagnostics", value = "Yes")                                                                                                                        
 twinData= umx_scale_wide_twin_data(data=twinData,varsToScale="wt",sep= "")                                                                                                 
 # Cut BMI column to form ordinal obesity variables                                                                                                                         
 obLevels   = c('normal', 'overweight', 'obese')                                                                                                                            
 cuts       = quantile(twinData[, "bmi1"], probs = c(.5, .2), na.rm = TRUE)                                                                                                 
 twinData$obese1=cut(twinData$bmi1,breaks=c(-Inf,cuts,Inf),labels=obLevels)                                                                                                 
 twinData$obese2=cut(twinData$bmi2,breaks=c(-Inf,cuts,Inf),labels=obLevels)                                                                                                 
 # Make the ordinal variables into mxFactors                                                                                                                                
 ordDVs = c("obese1", "obese2")                                                                                                                                             
 twinData[, ordDVs] = umxFactor(twinData[, ordDVs])                                                                                                                         
 mzData = twinData[twinData$zygosity %in% "MZFF",]                                                                                                                          
 dzData = twinData[twinData$zygosity %in% "DZFF",]
                                                                                                                          
 # MODIFICATION HERE: Use all the data so we can really exercise                                                                                                                               
 #mzData = mzData[1:80,] # just top 80 so example runs in a couple of secs                                                                                                  
 #dzData = dzData[1:80,]                                                                                                                                                    
                                                                                                                                                                            
 m1 = umxACE(selDVs= c("wt","obese"), dzData= dzData, mzData= mzData, sep='')                                                                                               
 
 # MODIFICATION HERE: toggle rowwise parallel                                                                                                                                                                           
 #m1$MZ$fitfunction$rowwiseParallel = FALSE                                                                                                                                 
 #m1$DZ$fitfunction$rowwiseParallel = FALSE                                                                                                                                 
                                                                                                                                                                            
 umx_set_cores(1); m2 = umxConfint(m1, run=TRUE) 
# MODIFICATION HERE: use 20 cores                                                                                                                       
 umx_set_cores(20); m3 = umxConfint(m1, run=TRUE)                                                                                                                           
 umx_time(c(m1, m2,m3))    
```

**Comment by RMKirkpatrick** [comment]
@khusmann Oh, OK.  I overlooked the fact that you ran his exact same code.

**Comment by khusmann** [comment]
Ok, so I decided to run with a larger sample to see how things scale. I've duplicated the rows of mzData and dzData 500 times in @tbates example so now my rows are 616000 and 375500 respectively. Running with rowwiseParallel=T gives:

```
ACE: 06.14 seconds, 
ACE: 42.13 seconds(Δ: 455.988), 7521.648%
ACE: 00.35 seconds(Δ: 534.213), 8794.836%
```

@tbates looks like the there's an issue with the time formatting string. Running `umx_time(c(m1,m2,m3), formatStr="std")` instead gives:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:06.14, 
ACE: Wall clock time (HH:MM:SS.hh): 00:07:42.13(Δ: 455.988), 7521.648%
ACE: Wall clock time (HH:MM:SS.hh): 00:09:00.35(Δ: 534.213), 8794.836%
```

I'm not sure rowwiseParallel is working properly?

**Comment by jpritikin** [comment]
> I'm not sure rowwiseParallel is working properly?

Does this model have any definition variables? If not, then the number of rows doesn't matter because the dataset will be automatically transformed into sufficient statistics.

**Comment by khusmann** [comment]
> Does this model have any definition variables?

Not that I know of... this is just a simple ACE model, right @tbates ?

>  If not, then the number of rows doesn't matter because the dataset will be automatically transformed into sufficient statistics.

So it just fits on the cov matrix? Wouldn't we expect execution time to be independent of the # rows then? (plus some headroom to calculate the cov)

In this example increasing the # of rows 500x increases execution time from 6 seconds to 7 *minutes*, so it looks to me like it's very dependent on the # of rows.

> Unique ordinal patterns will also force per-row evaluation.

What are unique ordinal patterns? Does this apply to the simple ACE model?

**Comment by RMKirkpatrick** [comment]
I think I can explain why Tim's model runs more slowly with 20 threads when `rowwiseParallel=FALSE`: the model only has 11 free parameters.  Thus, SLSQP only has 11 gradient elements to calculate numerically.  That is, SLSQP has no use for 9 of the 20 threads.  Those 9 threads' just increase the overhead, without increasing speed.

**Comment by RMKirkpatrick** [comment]
Wait, hang on.  In @tbates ' and @khusmann 's syntax above, `m1` doesn't estimate confidence intervals!  It makes no sense to use it as a basis for comparison to `m2` and `m3`, which (I guess) do nothing but estimate confidence intervals, which is usually the most time-consuming step.

**Comment by RMKirkpatrick** [comment]
So with that in mind, the results @tbates and @khusmann report make sense if we assume 20 threads truly is overkill in @khusmann 's case, right?

**Comment by khusmann** [comment]
@RMKirkpatrick Here's a few runs toggling rowwiseParallel and different # threads:

The overhead of having 9 extra threads doesn't seem to impact the performance when rowwiseParallel=F, but everything is bad when rowwiseParallel=T. 

Running with rowwiseParallel=F and 11 threads:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:06.15, 
ACE: Wall clock time (HH:MM:SS.hh): 00:07:44.90(Δ: 458.752), 7549.317%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:21.80(Δ: 15.65), 354.131%
```

Running with rowwiseParallel=T and 11 threads:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:04.46, 
ACE: Wall clock time (HH:MM:SS.hh): 00:04:14.27(Δ: 249.806), 5689.73%
ACE: Wall clock time (HH:MM:SS.hh): 00:04:25.04(Δ: 260.579), 5930.775%
```

Running with rowwiseParallel=F and 20 threads:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:04.36, 
ACE: Wall clock time (HH:MM:SS.hh): 00:04:15.45(Δ: 251.088), 5847.88%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:21.41(Δ: 17.045), 490.203%
```

Running with rowwiseParallel=T and 20 threads:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:06.14, 
ACE: Wall clock time (HH:MM:SS.hh): 00:07:42.13(Δ: 455.988), 7521.648%
ACE: Wall clock time (HH:MM:SS.hh): 00:09:00.35(Δ: 534.213), 8794.836%
```

Let me see if I understand what's going on here:

* rowwiseParallel=T parallelizes calculations of the objective function into jobs across rows, which is good in general for fitting models

* but for calculating confidence intervals, the bottleneck isn't calculating the objective function; it's calculating the gradient for each free parameter. So to make this step the most efficient you would ideally have a core for each free parameter. But rowwiseParallel=T prevents this. so we set rowwiseParallel=F to allow the parallelization to happen over each free parameter instead of splitting jobs across rows.

So long of it short, when I am running a large model (esp with many free params), I should:

* keep rowwiseParallel=T when fitting the model

* set rowwiseParallel=F when I calculate the confidence intervals, and make sure # threads <= # free parameters (although having extra threads doesn't seem to hurt at all here)

Am I on the right track now? Thanks for helping me understand here!

**Comment by RMKirkpatrick** [comment]
[mxCI_internals_180806_cropped.pdf](https://github.com/OpenMx/OpenMx/files/5059239/mxCI_internals_180806_cropped.pdf)
You're correct about this part,

> rowwiseParallel=T parallelizes calculations of the objective function into jobs across rows, which is good in general for fitting models

but I think that's about it.  Take a look at the attachment to this post.  During a confidence-limit search, evaluating the -2logL at the current values of the free parameters, as well as its gradient, always have to happen.  In the unconstrained representation, the objective function is subject to a quadratic penalty involving the -2logL; in the constrained optimization, there is a constraint placed on how much the -2logL can change from its value at the MLE.  In the unconstrained representation, the gradient of the objective depends upon the gradient of the -2logL; in the constrained representation, the constraint Jacobian depends upon the gradient of the -2logL.  

My suspicion is that calculating the -2logL in a row-wise parallel fashion doesn't work correctly for some reason when the -2logL has to be calculated as part of the constraint.  N.B. (1) that SLSQP defaults to using the constrained representation (whereas NPSOL and CSOLNP default to the unconstrained representation), (2) that `umxConfInt()` defaults to using SLSQP, and (3) only SLSQP can parallelize its gradient calculation.  What happens if you pass `optimizer="CSOLNP"` to `umxConfInt()` in your script?  I expect that with `rowwiseParallel=FALSE`, using multiple threads will be no faster than using 1 thread.  What I don't know is what happens with `rowwiseParallel=TRUE`.

**Comment by RMKirkpatrick** [comment]
BTW, _ceteris paribus_, we would expect better performance from SLSQP with `rowwiseParallel=FALSE`, because parallelizing over gradient elements is at a higher level of granularity than parallelizing over rows.

**Comment by khusmann** [comment]
@RMKirkpatrick 

Gotcha. This is starting to make sense.  With `optimizer="CSOLNP"`, and rowwiseParallel=F, I get:

```
> umx_set_cores(1); m3 = umxConfint(m1, run=TRUE, optimizer="CSOLNP"); umx_time(m3)
ACE: 04.57 seconds
> umx_set_cores(11); m3 = umxConfint(m1, run=TRUE, optimizer="CSOLNP"); umx_time(m3)
ACE: 05.40 seconds
> umx_set_cores(20); m3 = umxConfint(m1, run=TRUE, optimizer="CSOLNP"); umx_time(m3)
ACE: 05.78 seconds
```

Results are approximately identical to the above when rowwiseParallel=T.

Damn, that's fast. This is because we're approximating the equality constraint with a quadratic, then? So it's not going to be as good of an estimate as SLSQP?

**Comment by RMKirkpatrick** [comment]
> Damn, that's fast. This is because we're approximating the equality constraint with a quadratic, then?

That's part of it, yes.  There's also the fact that CSOLNP defaults to using a forward-difference approximation of the gradient, whereas SLSQP defaults to using a central-difference approximation.  I don't remember if CSOLNP uses Richardson extrapolation, but SLSQP definitely does.

> So it's not going to be as good of an estimate as SLSQP?

Well, the constrained representation is theoretically superior to the unconstrained one.  But, the optimizers' defaults are what they are because @jpritikin 's testing from a few years ago indicated that CSOLNP and NPSOL get better results without the explicit constraint, whereas SLSQP gets better results with it.  How similar are the CIs you get from SLSQP & CSOLNP?

**Comment by khusmann** [comment]
Gotcha. This helps. Thanks for all the explanations.

> BTW, ceteris paribus, we would expect better performance from SLSQP with rowwiseParallel=FALSE

Based on all this, would it make sense for rowwiseParallel to automatically turn off when using SLSQP then? Should that be the feature request to come out of this thread?

> How similar are the CIs you get from SLSQP & CSOLNP?

Can't tell, because I get all NAs with both (see @tbates output above).

**Comment by jpritikin** [comment]
> Based on all this, would it make sense for rowwiseParallel to automatically turn off when using SLSQP then?

Sometimes, but it's not always the best approach.

> Should that be the feature request to come out of this thread?

It's clearly too confusing to understand how OpenMx uses multiple threads. For starters, we need better diagnostics to make it easily visible what is going on.

**Comment by khusmann** [comment]
> For starters, we need better diagnostics to make it easily visible what is going on.

That would be quite nice. Esp paired with large examples users can run to get the hang of it.

**Comment by RMKirkpatrick** [comment]
> Can't tell, because I get all NAs with both (see @tbates output above).

Use `summary()` with argument `verbose=TRUE`.

**Comment by khusmann** [comment]
> Sometimes, but it's not always the best approach.

I see. I just tried a big model with many free parameters, and it looks like rowwiseParallel=T with 11 cores iterates faster than rowwiseParallel=F with 20 cores for SLSQP in that context.

> For starters, we need better diagnostics

How about adding "rate # iterations / sec" to the periodic optimizer output?

> How similar are the CIs you get from SLSQP & CSOLNP?
> Use summary() with argument verbose=TRUE.

Vaguely similar?

CSOLNP:
```
CI details:
        parameter  side     value     fit              diagnostic
1  top.a_std[1,1] lower 0.5312219 6579112 alpha level not reached
2  top.a_std[1,1] upper 0.5312219 6579112 alpha level not reached
3  top.a_std[2,1] lower 0.4408580 6579112 alpha level not reached
4  top.a_std[2,1] upper 0.4408580 6579112 alpha level not reached
5  top.a_std[1,2] lower 0.0000000 6579112 alpha level not reached
6  top.a_std[1,2] upper 0.0000000 6579112 alpha level not reached
7  top.a_std[2,2] lower 0.3354529 6579112 alpha level not reached
8  top.a_std[2,2] upper 0.3354529 6579112 alpha level not reached
9  top.c_std[1,1] lower 0.6595847 6579112 alpha level not reached
10 top.c_std[1,1] upper 0.6595847 6579112 alpha level not reached
11 top.c_std[2,1] lower 0.4833225 6579112 alpha level not reached
12 top.c_std[2,1] upper 0.4833225 6579112 alpha level not reached
13 top.c_std[1,2] lower 0.0000000 6579112 alpha level not reached
14 top.c_std[1,2] upper 0.0000000 6579112 alpha level not reached
15 top.c_std[2,2] lower 0.3354283 6579112 alpha level not reached
16 top.c_std[2,2] upper 0.3354283 6579112 alpha level not reached
17 top.e_std[1,1] lower 0.5317437 6579112 alpha level not reached
18 top.e_std[1,1] upper 0.5317437 6579112 alpha level not reached
19 top.e_std[2,1] lower 0.4841587 6579112 alpha level not reached
20 top.e_std[2,1] upper 0.4841587 6579112 alpha level not reached
21 top.e_std[1,2] lower 0.0000000 6579112 alpha level not reached
22 top.e_std[1,2] upper 0.0000000 6579112 alpha level not reached
23 top.e_std[2,2] lower 0.3355489 6579112 alpha level not reached
24 top.e_std[2,2] upper 0.3355489 6579112 alpha level not reached
```

SLSQP:
```
CI details:
        parameter  side     value     fit              diagnostic
1  top.a_std[1,1] lower 0.4930153 6601944 alpha level not reached
2  top.a_std[1,1] upper 0.5022897 6586141 alpha level not reached
3  top.a_std[2,1] lower 0.4583781 6599041 alpha level not reached
4  top.a_std[2,1] upper 0.4676762 6610976 alpha level not reached
5  top.a_std[1,2] lower 0.0000000 6511176 alpha level not reached
6  top.a_std[1,2] upper 0.0000000 6505911 alpha level not reached
7  top.a_std[2,2] lower 0.3313577 6724607 alpha level not reached
8  top.a_std[2,2] upper 0.3120349 6515231 alpha level not reached
9  top.c_std[1,1] lower 0.6333872 6627975 alpha level not reached
10 top.c_std[1,1] upper 0.6664507 6532949 alpha level not reached
11 top.c_std[2,1] lower 0.5297226 6523332 alpha level not reached
12 top.c_std[2,1] upper 0.4957158 6574131 alpha level not reached
13 top.c_std[1,2] lower 0.0000000 6630008 alpha level not reached
14 top.c_std[1,2] upper 0.0000000 6607864 alpha level not reached
15 top.c_std[2,2] lower 0.3143214 6567496 alpha level not reached
16 top.c_std[2,2] upper 0.3320770 6721446 alpha level not reached
17 top.e_std[1,1] lower 0.5649144 6597668 alpha level not reached
18 top.e_std[1,1] upper 0.5281778 6619182 alpha level not reached
19 top.e_std[2,1] lower 0.4600057 6652117 alpha level not reached
20 top.e_std[2,1] upper 0.4595384 6628532 alpha level not reached
21 top.e_std[1,2] lower 0.0000000 6683969 alpha level not reached
22 top.e_std[1,2] upper 0.0000000 6629989 alpha level not reached
23 top.e_std[2,2] lower 0.3336665 6681149 alpha level not reached
24 top.e_std[2,2] upper 0.3223011 6495126 alpha level not reached
```

**Comment by mcneale** [comment]
Alpha level not reached is reasonable for something hitting a lower bound of zero.  It does not seem reasonable for a search for an upper confidence interval, especially as estimate of upper CI is sometimes lower than the estimate of the lower CI.  

Either this is a particularly gnarly problem, or maybe the model is not identified, or there really is a bug.  Developers are often and naturally reluctant to admit the last of these, because it means extra unplanned work :(.

**Comment by RMKirkpatrick** [comment]
> Alpha level not reached is reasonable for something hitting a lower bound of zero. It does not seem reasonable for a search for an upper confidence interval, especially as estimate of upper CI is sometimes lower than the estimate of the lower CI.

PROTIP: if the lower and upper confidence limit are both zero, then the reference quantity is quite likely to be fixed to zero.

**Comment by RMKirkpatrick** [comment]
@khusmann I'm curious, what would 3.841 plus your -2logL at the MLE be?

**Comment by khusmann** [comment]
@RMKirkpatrick where does the 3.841 come from? My -2logL was 6583567

```
Model Statistics: 
               |  Parameters  |  Degrees of Freedom  |  Fit (-2lnL units)
       Model:             11                3872489               6583567
```

**Comment by jpritikin** [comment]
It wasn't as easy as I hope, but I think I got it sorted. Try 97dfee160c58e

**Comment by mcneale** [comment]
> > Alpha level not reached is reasonable for something hitting a lower bound of zero. It does not seem reasonable for a search for an upper confidence interval, especially as estimate of upper CI is sometimes lower than the estimate of the lower CI.

> PROTIP: if the lower and upper confidence limit are both zero, then the reference quantity is quite likely to be fixed to zero.

But that is NOT the case here:

7  top.a_std[2,2] lower 0.3313577 6724607 alpha level not reached
8  top.a_std[2,2] upper 0.3120349 6515231 alpha level not reached

**Comment by khusmann** [comment]
> Parallel evaluation of WLS fitfunction wasn't as easy as I hope, but I think I got it sorted. Try 97dfee1

@jpritikin this is great, thank you!

Sorry for my ignorance, but I still have a big model (around 100 free params) I'm trying to run that looks like it's only using one core for the confidence interval calculations (using SLSQP on a supermodel with multigroup fitfunction with ordinal submodels with ML fitfunctions), even when I set rowwiseParallel=F. With diagnostics it looks like it is spawning threads properly and rowwiseParallel=0 as it should:

```
create 20 FitContext for parallel processing; OpenMP user=0
dsc_model_none.fitfunction: rowwiseParallel = 0
dsc_model_both.fitfunction: rowwiseParallel = 0
dsc_model_decoding.fitfunction: rowwiseParallel = 0
dsc_model_comprehension.fitfunction: rowwiseParallel = 0
dsc_model_none.fitfunction: rowwiseParallel = 0
dsc_model_both.fitfunction: rowwiseParallel = 0
dsc_model_decoding.fitfunction: rowwiseParallel = 0
dsc_model_comprehension.fitfunction: rowwiseParallel = 0
dsc_model_none.fitfunction: rowwiseParallel = 0
dsc_model_both.fitfunction: rowwiseParallel = 0
dsc_model_decoding.fitfunction: rowwiseParallel = 0
dsc_model_comprehension.fitfunction: rowwiseParallel = 0
dsc_model_none.fitfunction: rowwiseParallel = 0
dsc_model_both.fitfunction: rowwiseParallel = 0
dsc_model_decoding.fitfunction: rowwiseParallel = 0
dsc_model_comprehension.fitfunction: rowwiseParallel = 0
...
[15/1] MxComputeGradientDescent(SLSQP) evaluations 15268 fit -0.413894 change 4.261e-05
```

When I add CIs to an already run model, I get a flurry of activity on all cores for a moment (when it presumably is re-finding the MLE and discovering it's already there) but then once it starts calculating the confidence intervals I just have 100% on one core. Am I missing something obvious here or is it worth it to try to whittle this down to a MWE? Thanks!

**Comment by RMKirkpatrick** [comment]
> @RMKirkpatrick where does the 3.841 come from? My -2logL was 6583567

3.841 is the 0.95 quantile of a chi-square distribution on 1df.  The -2logL at a 95% confidence limit should be worse than its value at the MLE by 3.841.  I was wondering how badly the confidence-limit search missed its alpha level.

**Comment by mcneale** [comment]
> When I add CIs to an already run model, I get a flurry of activity on all cores for a moment (when it presumably is re-finding the MLE and discovering it's already there) but then once it starts calculating the confidence intervals I just have 100% on one core. Am I missing something obvious here or is it worth it to try to whittle this down to a MWE? Thanks!

I suspect that if rowwiseParallel is False (0) then: i) parallel isn't used to evaluate the fit function, and ii) the larger granularity of doing CIs in parallel is not working at all.  It seems likely that you would get better performance with the CIs with rowwiseParallel set to TRUE, but that the improvement might be less than if the different CIs had been done in parallel.

**Comment by khusmann** [comment]
> i) parallel isn't used to evaluate the fit function

Since I'm using SLSQP and rowwiseParallel = F, wouldn't we expect each core to be calculating a gradient?

> ii) the larger granularity of doing CIs in parallel is not working at all.

I see, so this goes back to @tbates 's OP

> It seems likely that you would get better performance with the CIs with rowwiseParallel set to TRUE, but that the improvement might be less than if the different CIs had been done in parallel.

Or maybe with my 20 cores I'd want to be able to run 5 CIs with rowwiseParallel on 4 cores each, or something like that?

**Comment by tbates** [comment]
With rowwiseParallel = FALSE and a larger data set, CI run times explode from 5.6 _seconds_  on 1 core to 3 MINUTES 31 secs on 4 cores!

So using multi-core, even with fewer cores than CIs  is much worse with rowwiseParallel = FALSE (for ML with SLSQP)

**Comment by khusmann** [comment]
@tbates Interesting. On my 20 core machine, rowwiseParallel=F is faster with 4 cores. (Keep reading, it gets weirder...)

```
m1$MZ$fitfunction$rowwiseParallel = TRUE
m1$DZ$fitfunction$rowwiseParallel = TRUE                                                                                                                                   

umx_set_cores(1); m2 = umxConfint(m1, run=TRUE) 
umx_set_cores(4); m3 = umxConfint(m1, run=TRUE)
umx_set_cores(8); m4 = umxConfint(m1, run=TRUE)
umx_set_cores(11); m5 = umxConfint(m1, run=TRUE)                                                                                     

m1$MZ$fitfunction$rowwiseParallel = FALSE                                                                                                                                  
m1$DZ$fitfunction$rowwiseParallel = FALSE                                                                                                                                  
                                                                                                                                                                            
umx_set_cores(1); mm2 = umxConfint(m1, run=TRUE)
 umx_set_cores(4); mm3 = umxConfint(m1, run=TRUE)
 umx_set_cores(8); mm4 = umxConfint(m1, run=TRUE)
 umx_set_cores(11); mm5 = umxConfint(m1, run=TRUE)
                                                                                                                                                                            
 umx_time(c(m2,m3,m4,m5), formatStr='std')                                                                                                                                  
 umx_time(c(mm2,mm3,mm4,mm5), formatStr='std')   
```

Results:

```
ACE: Wall clock time (HH:MM:SS.hh): 00:00:07.29, 
ACE: Wall clock time (HH:MM:SS.hh): 00:00:06.95(Δ: -0.344), 95.293%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:07.23(Δ: -0.064), 99.128%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:07.64(Δ: 0.342), 104.685%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:06.70, 
ACE: Wall clock time (HH:MM:SS.hh): 00:04:29.91(Δ: 263.207), 4023.338%
ACE: Wall clock time (HH:MM:SS.hh): 00:02:39.84(Δ: 153.131), 2382.558%
ACE: Wall clock time (HH:MM:SS.hh): 00:02:49.99(Δ: 163.291), 2534.001%
```

BUT look at what happens when I increase the number of rows...

```
mzData <- mzData[rep(seq_len(nrow(mzData)), 500), ]                                   
dzData <- mzData[rep(seq_len(nrow(dzData)), 500), ] 
```

Now rowwiseParallel=F becomes massively superior (down to 20 SECONDS). Does this happen for you @tbates? 

```
ACE: Wall clock time (HH:MM:SS.hh): 00:08:24.18, 
ACE: Wall clock time (HH:MM:SS.hh): 00:08:34.65(Δ: 10.478), 102.078%
ACE: Wall clock time (HH:MM:SS.hh): 00:08:32.18(Δ: 8.007), 101.588%
ACE: Wall clock time (HH:MM:SS.hh): 00:08:30.29(Δ: 6.115), 101.213%
ACE: Wall clock time (HH:MM:SS.hh): 00:08:26.47, 
ACE: Wall clock time (HH:MM:SS.hh): 00:00:40.92(Δ: -465.544), 8.081%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:29.04(Δ: -477.433), 5.734%
ACE: Wall clock time (HH:MM:SS.hh): 00:00:20.47(Δ: -485.998), 4.043%
```

**Comment by mcneale** [comment]
This is probably because OpenMx has some cunning algorithms to speed up raw data analysis.  If the pattern of missing data is the same as prior row, no need to recalculate & invert the expected covariance matrix.  If the data are the same as well, no need to do anything other than add the previous row's log-likelihood again.  Data are sorted to try to maximize such savings.  So the rep() approach won't be the same as, e.g., mvrnorm()'ing some new data rows.

**Comment by tbates** [comment]
So @khusmann, with `rowwiseParallel = TRUE` we are all seeing flat performance.

re larger data: You copied the mzData into the dzData - that will alter things :-)

But also @mcneale is right, copying to increase datasize won't slow the model down. In fact, without NAs, the data will be converted to a single cov matrix - should be zero effect of  rows...?).

`mxGenerateData` is the mx function to generate new data from a model (or data frame).

```R
m1big = mxGenerateData(m1, nrowsProportion =500, returnModel=TRUE)
m1big = mxRename(m1big, "bigACE")
```
(see also, however, #287 )

My guess is parallel CIs are not working, and `rowwiseParallel = FALSE` is doing something nasty to evaluation such that multi-core performance on CIs is slowed for any number of cores > 1, but that creating hundreds of duplicate rows makes the model ridiculously confident about the shape of whatever function it is evaluating to estimate CIs), overcoming this reduction in speed by making CI-computation essentially noise-free?

**Comment by jpritikin** [comment]
> Sorry for my ignorance, but I still have a big model (around 100 free params) I'm trying to run that looks like it's only using one core for the confidence interval calculations (using SLSQP on a supermodel with multigroup fitfunction with ordinal submodels with ML fitfunctions), even when I set rowwiseParallel=F. With diagnostics it looks like it is spawning threads properly and rowwiseParallel=0 as it should:

Yeah, I can confirm that something is going wrong. I'm going to add more diagnostics. Give me a few days to figure it out.

**Comment by khusmann** [comment]
> re larger data: You copied the mzData into the dzData - that will alter things :-)

Wooops! Good catch :-)

> mxGenerateData is the mx function to generate new data from a model (or data frame).

Ah, very useful, thanks!

> In fact, without NAs, the data will be converted to a single cov matrix - should be zero effect of rows...?

I would have also expected this... execution time to be independent of the # rows, plus some headroom to calculate the cov? Not what we're getting empirically. Very confusing. Eigen uses openmp too, right? I wonder if that might interact poorly with OpenMx's parallelization...?

> Yeah, I can confirm that something is going wrong. I'm going to add more diagnostics. Give me a few days to figure it out.

Hooray! I was starting to worry everything was working as it should and I was just going to have to get used to my large model requiring days to compute confints. Looking forward to seeing what you come up with. Thanks in advance for all your work on this @jpritikin (And thanks to everyone else on this thread too!)

**Comment by jpritikin** [comment]
> Thanks in advance for all your work on this @jpritikin

Thanks for testing! The support for parallel processing is hard to test automatically.

**Comment by jpritikin** [comment]
@khusmann I'm a bit embarrassed. The code was randomly deciding how many threads to use for the gradient. Try v2.17.3-161-g5a04c0e98

**Comment by jpritikin** [comment]
Hm, it looks like there is some non-thread-safe code in the confidence interval path. Investigating...

**Comment by jpritikin** [comment]
Just a note to look at #288 too.

**Comment by jpritikin** [comment]
In my experience, it's never just one bug. The first thing I fixed is 2a9304fbd44. However, CIs are still single threaded. For parallel speedup of CIs, we need parallel evaluation of the Jacobian. Working on that next.

**Comment by jpritikin** [comment]
OK, please try bbd5a7f21

With SLSQP, inequality constraints used in confidence intervals are now evaluated in parallel. You should see diagnostics related to this, if enabled. Look for JacobianGadget.

**Comment by khusmann** [comment]
@jpritikin ah, this is looking great! Thanks for your work! Now rowwiseParallel=F is faster than rowwiseParallel=T on my large model, as expected. (It's a little more than twice as fast, which I think makes sense if rowwiseParallel=T hits diminishing returns at 11 threads as we've seen before) Let me know when you've fixed whatever issues you're working on and I'll try it again.

Some thoughts --

1) Is this the debugging info from JackbianGadget or is this something else?

```[0] gradient took 11087.335347ms with 20 threads and 11082.506044ms with 19 threads```

It settles on 19 threads for me but I think it really should stick with the 20 threads, because I think the 5ms might just be noise. When it compares 19 and 18 later, the difference is more dramatic -- I think it might be settling on the 19 too soon.

2) Can we add a speed indicator to the regular output? something like:

```MxComputeGradientDescent(SLSQP) evaluations 1813 fit 318876 change -1.955e+04 40evals/sec```

That would help in my comparisons between rowwiseParallel=F and rowwiseParallel=T (right now i'm manually using a stopwatch) -- or maybe there's a different debug thing I need to enable to get this?

**Comment by jpritikin** [comment]
> Let me know when you've fixed whatever issues you're working on and I'll try it again

I think cb3ee7f94c is good to go, but something is broken with R on Travis. It will probably be fixed in a day or two.

> [0] gradient took 11087.335347ms with 20 threads and 11082.506044ms with 19 threads

It starts with "gradient" so we know this is for the gradient. During CI estimation, it will say "JacobianGadget" instead of "gradient"

> It settles on 19 threads for me but I think it really should stick with the 20 threads, because I think the 5ms might just be noise.

Maybe, but the way it works is that it takes 3 measurements of elapsed time for 19 and 20 threads interleaved, and then takes the median. So the elapsed time comparison should be pretty accurate. I don't think a difference of 5ms is worth worrying about.

> Can we add a speed indicator to the regular output? something like:

That's a neat idea. The problem is that we are limited to a single line and it's pretty full already. Could we replace the number of evaluations with eval/sec?

**Comment by khusmann** [comment]
> I think cb3ee7f is good to go, but something is broken with R on Travis

Ok great, I'll keep playing with it and let you know if anything else comes up.

> During CI estimation "JacobianGadget" instead of "gradient"

Sounds good.

> I don't think a difference of 5ms is worth worrying about.

Cool

> Could we replace the number of evaluations with eval/sec?

I like that idea. I only use the # of evals to get a sense of the rate, and a feel for how long it's been running... so what about eval/sec and total time elapsed? How about:

```[hh:mm:ss] MxComputeGradientDescent(SLSQP) 40evals/sec fit 318876 change -1.955e+04 ```

or something.

**Comment by RMKirkpatrick** [comment]
I ran the GREML test from models/nightly that uses gradient-descent and does confidence intervals.  I ran it with SLSQP, to compare running time for 1 vs. 2 threads.  With 2 threads, I can confirm that it was using both threads during the confidence-limit search, but I was unable to actually compare running times, since the 2-threaded run overheated my laptop.

**Comment by jpritikin** [comment]
> but I was unable to actually compare running times, since the 2-threaded run overheated my laptop.

Hm, which suggests that we should add an `mxOverheat(logical)` function to enable threads?

**Comment by khusmann** [comment]
@jpritikin ok, I'm seeing JacobianGadget now, and things are going much faster. Interestingly I have a pattern of all 20 cores working for like 30 seconds, then 1 core working for about 30 seconds, back and forth. Is this because the # of jobs is not a good multiple of my # of threads? Or is there a task it's doing that's not multithreaded?

**Comment by jpritikin** [comment]
> Or is there a task it's doing that's not multithreaded?

That would be my guess. Each CI is done one-at-a-time. I'd have to look at the code to speculate why there is a delay between CIs.

**Comment by khusmann** [comment]
> speculate why there is a delay between CIs

This is within one CI -- on my model one CI takes about an hour, I see it going between 20core and 1core usage every 30secs.

**Comment by jpritikin** [comment]
> This is within one CI

Oh, then what you're seeing is the optimizer switching between the line search (serial) and the inequality jacobian (parallel). There is certainly room for improvement, but it would take a lot of effort to reorg the code.

**Comment by khusmann** [comment]
> what you're seeing is the optimizer switching between the line search (serial) and the inequality jacobian (parallel)

gotcha, figured it was something like that. thanks for all your work!

**Comment by tbates** [comment]
FYI with `GIT v2.18.1-12-gc9d88fbd`  on a 4 core iMac, `rowwiseParallel = "No"` gives ~50% speed up going from 1-4 cores. Meanwhile the default `rowwiseParallel = "Yes"` get a smidgeon _worse_.

So:
1. If the CI computation fundamentally can't be parallelised with rowwise on, the default for `omxParallelCI` should be to turn `rowwiseParallel = "No"` in all models and submodels before setting off (and toggle back to old value after success).
2. Feels like there should still be room for a 2x improvement for parallel.


```R
mxOption(key="Parallel diagnostics", value = "Yes") 
data(twinData)
twinData= umx_scale_wide_twin_data(data=twinData,varsToScale="wt",sep= "")
# Cut BMI column to form ordinal obesity variables
obLevels   = c('normal', 'overweight', 'obese')
cuts       = quantile(twinData[, "bmi1"], probs = c(.5, .2), na.rm = TRUE)
twinData$obese1=cut(twinData$bmi1,breaks=c(-Inf,cuts,Inf),labels=obLevels)
twinData$obese2=cut(twinData$bmi2,breaks=c(-Inf,cuts,Inf),labels=obLevels)
# Make the ordinal variables into mxFactors
ordDVs = c("obese1", "obese2")
twinData[, ordDVs] = umxFactor(twinData[, ordDVs])
mzData = twinData[twinData$zygosity %in% "MZFF",] 
dzData = twinData[twinData$zygosity %in% "DZFF",]
m1 = umxACE(selDVs= c("wt","obese"), dzData= dzData, mzData= mzData, sep='')

m1$MZ$fitfunction$rowwiseParallel = FALSE
m1$DZ$fitfunction$rowwiseParallel = FALSE

umx_set_cores(1); m2 = umxConfint(m1, run=TRUE)
umx_set_cores(4); m3 = umxConfint(m1, run=TRUE)

umx_time(c(m2,m3))

# rowwiseParallel OFF
# ACE 1core: 06.14 sec , 
# ACE 4core: 02.94 sec (Δ: -3.2), 47.929%

# rowwiseParallel ON
# ACE 1core: 06.43 sec , 
# ACE 4core: 06.56 sec (Δ: 0.131), 102.031%

mxOption(key="Parallel diagnostics", value = "No")

**Comment by jpritikin** [comment]
> Feels like there should still be room for a 2x improvement for parallel.

As currently implemented, with `rowwiseParallel=FALSE`, the speedup depends on the number of parameters. The more parameters in your model, the more that parallel helps.

**Comment by jpritikin** [comment]
> If the CI computation fundamentally can't be parallelised with rowwise on

If you have a model where summary stats cannot be used (i.e., with lots of unique ordinal patterns or definition variables) then `rowwiseParallel=TRUE` should help when fitting and with CIs.

**Comment by tbates** [comment]
> > Feels like there should still be room for a 2x improvement for parallel.

> speedup depends on the number of parameters.

With (in this case) 12 parameters/24 intervals to compute, shouldn't 4 cores be saturated?

**Comment by jpritikin** [comment]
> With (in this case) 12 parameters/24 intervals to compute, shouldn't 4 cores be saturated?

It depends how long it takes to evaluate your model. That's why there are parallel diagnostics, to show you the amount of time conditional on the number of threads.

With a heavy enough model, 4 parameters could saturate 4 cores.

**Comment by jpritikin** [comment]
Note: When `rowwiseParallel` is enabled then every fit evaluation benefits. When `rowwiseParallel` is disabled, only the gradient benefits from parallel. The line search is serial.

**Comment by tbates** [comment]
In response to this issue: 
1. WLS can now duplicate
2. Smart default for `rowwiseParallel` implemented,
3. Choice of thread number is empirical rather than random
4. Non-thread-safe code in the confidence interval path was removed (?)

Maybe close? If desired could make a separate issue for making CIs parallel, parallelising the line search, and maybe making parallel not just per-parameter but per model solution, getting multi-chip (not just core) working.

**Comment by mcneale** [comment]
Multiprocessor systems seem a bit like a multi-node cluster inside a single box.  The difference is that a multiprocessor system can access the same memory, whereas multi-node approaches each have their own.  For some large memory problems, multiprocessor might have an advantage by accessing the same RAM.  The programming does not seem to get any easier with parallel for multiprocessor - reports suggest that it's more tricky.  We are, not uncommonly, faced with trying to guess the way in which the world is going to change.

**Comment by tbates** [comment]
[Fun watch](https://www.youtube.com/watch?v=nWTvXbQHwWs) - almost 3-hours with the creator of LLVM and Swift, and now RISC-V (open-instruction set) SiFive fabless chip company. Gets into abstractions to enable parallelisation (from around [5182 seconds](https://www.youtube.com/watch?v=nWTvXbQHwWs&t=5182s) ) :-)
